### ISSUE TITLE
Support -P / --progress with progress-bar and ETA like odeke-em/drive pull

### ISSUE THREAD

**Issue by devurandom** [ncw/rclone#2347]
#### What is the problem you are having with rclone?

It would be nice to have `-P` / `--progress` command line flags that trigger an output mode similar to that of odeke-em/drive, which displays a progress bar (bytes transferred / bytes total) with an ETA for the `drive pull` command.

"Mockup":
```
$ rclone copy -P googledrive:SOME/PATH/ ./
123K / 2.1G [===>                 ] 1h 10m
```

An implementation probably would not require [ncurses](https://www.gnu.org/software/ncurses/) or [termbox](https://github.com/nsf/termbox). I assume it would be enough to take care to not end the progress-bar line with `\n`, but `\r` instead.

#### What is your rclone version (eg output from `rclone -V`)

```
rclone v1.41
- os/arch: linux/amd64
- go version: go1.10.1
```

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)

Linux 4.16, x86_64

####  Which cloud storage system are you using? (eg Google Drive)

WebDAV / NextCloud

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

`rclone copy googledrive:SOME/PATH/ ./`

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

Logs nothing by default and `INFO` log messages with `-v`, neither is what I would like to see.

**Comment by devurandom** [comment]
What's particularly important to me is to see the amount of data that will be transferred in total, and the estimated time it will take to do that. I am not so much interested in statistics for individual files, because in my use-case, most files are only a few megabyte in size. Thus stats displays information that I am not interested in, while omitting what is.

**Comment by ncw** [comment]
rclone doesn't know the amount of stuff that needs to be transferred in total without scanning the source and destination first.  Normally rclone does this in parallel with the transfers to speed things up and save memory.

So what rclone could do is the equivalent of do a `--dry-run` sync writing each file that needs transferring to a file, then do a sync with `--files-from` just to transfer those files.

**Comment by ncw** [comment]
I made a sort of proof of concept of how the stats and the logging would combine.  So the lines which start `Line` are the stats and the others are scrolling log lines.

[![asciicast](https://asciinema.org/a/189216.png)](https://asciinema.org/a/189216)

What do you think?

**Comment by 19wolf** [comment]
I like this proof of concept, but I would also love a "total bytes transferred" or similar locked at the bottom

**Comment by ncw** [comment]
OK here is attempt number 2 within rclone which I think is getting there...

[![asciicast](https://asciinema.org/a/189525.png)](https://asciinema.org/a/189525)

**Comment by komposure** [comment]
@ncw How's this going? Last result looks very good!

**Comment by ncw** [comment]
I've been on vacation and conferences but will get back to it shortly!

**Comment by ncw** [comment]
OK here is the next iteration. This has whole operation stats and supports --stats-one-line along with --progress.  I tried to make `--progress --one-line` look like rsync's `progress2` equivalent.

https://beta.rclone.org/branch/v1.42-126-g0475830a-fix-2347-progress/ (uploaded in 15-30 mins)

I haven't implemented @ewwink suggestion about switching the order of the transferring files and the transfer stats yet.

**Comment by ncw** [comment]
That one didn't build, try this one...

https://beta.rclone.org/branch/v1.42-128-gebf185ec-fix-2347-progress/ (uploaded in 15-30 mins)

**Comment by ncw** [comment]
3rd time lucky - this did build :-)

https://beta.rclone.org/branch/v1.42-128-g816d3ceb-fix-2347-progress/

**Comment by ncw** [comment]
I have merged that in

https://beta.rclone.org/v1.42-131-g1cccfa73/ (uploaded in 15-30 mins)

which will be in [the latest beta](https://beta.rclone.org) and v1.34

I expect that it will get  tweaked as time goes by!

**Comment by 19wolf** [comment]
I like it, though I wish I could change how often it updates, anywhere from the current .5s to maybe 5s?
If I'm uploading a large amount of data (2GB+) I don't need an update every half second if it's going to take 5 minutes, every 2-5 seconds should suffice.

**Comment by kpcyrd** [comment]
Allowing to override this seems like a good idea, but 0.5s seems like a sane default for a casual `rclone copy -vP foo: bar:` call.

For anybody else wondering, this feature is available by installing the [rclone-git](https://aur.archlinux.org/packages/rclone-git) package until a new version is released. To migrate existing scripts remove the `--stats*` flags and add `-P`.

**Comment by kpcyrd** [comment]
@ncw I think I found a bug in the current beta (`v1.42.r133.g174ca229`):

If you transfer large files you're going to notice that the transferred line doesn't know the full size (always reports 100%), while the transferring line (the progress indicator) knows the final size:

```
Transferred:      937.001M / 937.001 MBytes, 100%, 8.167 MBytes/s, ETA -
Errors:                 0
Checks:                 0 / 0, 0%
Transferred:            2 / 2, 100%
Elapsed time:     1m54.7s
Transferring:
 * file.bin: 32% /2.850G, 6.933M/s, 4m45s
```

**Comment by ncw** [comment]
@19wolf wrote:
> I like it, though I wish I could change how often it updates, anywhere from the current .5s to maybe 5s?
If I'm uploading a large amount of data (2GB+) I don't need an update every half second if it's going to take 5 minutes, every 2-5 seconds should suffice.

Just use the `--stats 5s` flag to change the time.

Do you think 0.5s is too fast?

@kpcyrd wrote:
> If you transfer large files you're going to notice that the transferred line doesn't know the full size (always reports 100%), while the transferring line (the progress indicator) knows the final size:

Ah I see what you mean.  The Transferred: line is only taking account of things in the queue, not things being transferred.

I've attempted to fix that here - can you give it test please?

https://beta.rclone.org/branch/v1.42-143-g799961a9-fix-stats/ (uploaded in 15-30 mins)

**Comment by kpcyrd** [comment]
@ncw That fixed it, thanks! I've noticed the speed and ETA don't match while transferring a single file, but that's probably harder to fix.

**Comment by ncw** [comment]
@kpcyrd 
OK I've merged that!

> I've noticed the speed and ETA don't match while transferring a single file, but that's probably harder to fix.

I noticed that too...  I had a brainwave and have managed to fix it

Those fixes will be in master

https://beta.rclone.org/v1.42-147-gde6ec805/ (uploaded in 15-30 mins)

and [the latest beta](https://beta.rclone.org) from now on.

The speeds may not match exactly since one is an average over the whole transfer and the other is an average over the last 16 seconds or so but it should much better now!

**Comment by dannymichel** [comment]
How would i install the beta version with  progress on MacOS?

**Comment by ncw** [comment]
@dannymichel those fixes are in 1.43.1 so you can get that from rclone.org/downloads

**Comment by dannymichel** [comment]
oh thanks

**Comment by dannymichel** [comment]
Ok cool so it's just `rclone -P copy source dest`

**Comment by ncw** [comment]
That's it!

**Comment by gauteh** [comment]
If I do:
````
echo "syncing..."
rclone --stats-one-line -P something something:asdf
echo "syncing something else.."
```
the rclone status line does not terminate with a new line, so the second "syncing something else" is appened to the rclone status rather than printed on a new line.

**Comment by gauteh** [comment]
Nick Craig-Wood writes on October 4, 2018 9:24:
> @gauteh I've just fixed this - find the fix in [the latest beta](https://beta.rclone.org) (in 15-30 mins) - thanks for reporting.

Thanks :)

**Comment by WiliTest** [comment]
Excellent! It looks like this : 
rclone sync --dry-run "C:\Users\Me\Desktop\test"  "test drive:test-me"  --log-file %logpath%" --log-level INFO --delete-after --copy-links --drive-skip-gdocs **--stats-one-line -P --stats 2s**
